File: C++/프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<bits/stdc++.h>
#include<vector>
using namespace std;
int visited[104][104];
queue<pair<int,int>> q;
int dy[4] = {1,0,-1,0};
int dx[4] = {0,1,0,-1};
void bfs(int a,int b,int n, int m,vector<vector<int>>& maps){
    q.push({a, b});
    visited[a][b]=1;
    while(!q.empty()){
        int y=q.front().first;
        int x=q.front().second;
        q.pop();
        for(int i=0;i<4;i++){
            int ny=y+dy[i];
            int nx=x+dx[i];
            if(ny < 0 or nx < 0 or ny>=n or nx>=m) continue;
            if(visited[ny][nx]!=-1) continue;
            if(maps[ny][nx]==0) continue;
            q.push({ny,nx});
            visited[ny][nx]=visited[y][x]+1;
        }
    }
}

int solution(vector<vector<int> > maps)
{
    //visited 처음 초기화???
    for(int i=0;i<=100;i++){
        for(int j=0;j<=100;j++){
            visited[i][j]=-1;
        }
    }
    
    bfs(0,0,maps.size(),maps[0].size(),maps);
    int answer = visited[maps.size()-1][maps[0].size()-1];
    
    return answer;
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
// 최단거리 bfs
let dy = [1, 0, -1,  0]
let dx = [0, 1,  0, -1]

function solution(maps) {
    var answer = 0;

    let visited = [];
    
    for(let i = 0; i < maps.length; i++){
        visited[i] = [];
        for(let j = 0; j < maps[0].length; j++){
            visited[i][j] = 0;
        }
    }
    
    console.log(visited);
    
    let q = [];
    
    q.push([0, 0, 1])
    visited[0][0] = 1;
    
    while(q.length != 0){
        let y = q[0][0];
        let x = q[0][1];
        let dis = q[0][2];
        q.shift(); //시간 복잡도가 N이긴 해서 맞긴 않고 사파
        
        for(let i = 0; i < 4; i++){
            let ny = y + dy[i];
            let nx = x + dx[i];
            
            if(ny < 0 || ny >= maps.length || nx < 0 || nx >= maps[0].length) continue;
            if(maps[ny][nx] == 0) continue;
            if(visited[ny][nx] != 0) continue;
            
            visited[ny][nx] = dis + 1;
            q.push([ny, nx, dis + 1]);
        }
    
    
    }
    answer = visited[maps.length - 1][maps[0].length - 1]
    if(answer == 0) answer = -1;
    return answer;
}